EIPS/eip-5169.md

```diff
@@ -0,0 +1,191 @@
+---
+eip: 5169
+title: Client Script URI for Token Contracts
+description: Add a scriptURI to point to point to an executable script associated with the functionality of the token.
+author: James (@JamesSmartCell), Weiwu (@weiwu-zhang)
+discussions-to: https://ethereum-magicians.org/t/eip-5133-delaying-difficulty-bomb-to-mid-september-2022/9622
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-05-03
+requires:
+---
+
+### Abstract
+This ERC is a contract interface that adds a `scriptURI()` function for locating executable scripts associated with the token.
+
+
+### Motivation
+
+Often, Smart Contract authors want to provide some user functionality to their tokens through client scripts. The idea is made popular with function-rich NFTs.
+
+It's important that a token's client script is linked with the token's contract, since the client script carry out trusted tasks such as creating transactions for the user.
+
+Users can be sure they are using the correct script through the contract author packaging a URI to an official script, made available with a call to the token contract itself.
+
+This ERC proposes adding a `scriptURI`, a structure containing URIs to download the client script. It can be a link to IPFS, GitHub, a cloud provider, etc.
+
+Each token contract has one `scriptURI`  function to return the download URI to a client script.
+
+Concretely each element in the array contains a URI to download the script itself.
+
+The script provides a client-side executable to the hosting token. Examples of such a script could be:
+
+- A 'miniDapp', which is a cut-down dapp tailored for a single token.
+- A 'TokenScript' which provides [T.I.P.S.](https://tokenscript.org/TIPS.html) from a browser wallet.
+- An extension that is downloadable to the hardware wallet with an extension framework, such as Ledger.
```

Comment:
```suggestion
Often, smart contract authors want to provide some user functionality to their tokens through client scripts. The idea is made popular with function-rich NFTs. It's important that a token's contract is linked to its client script, since the client script may carry out trusted tasks such as creating transactions for the user.

This EIP allows users to be sure they are using the correct script through the contract by providing a URI to an official script, made available with a call to the token contract itself (`scriptURI`). This URI can be any RFC 3986-compliant URI, such as a link to an IPFS multihash, GitHub gist, or a cloud storage provider. Each contract implementing this EIP  implements a `scriptURI` function which returns the download URI to a client script. The script provides a client-side executable to the hosting token. Examples of such a script could be:
- A 'miniDapp', which is a cut-down DApp tailored for a single token.
- A 'TokenScript' which provides TIPS from a browser wallet.
- An extension that is downloadable to the hardware wallet with an extension framework, such as Ledger.
```